src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs

```diff
@@ -700,32 +700,86 @@ public void ExecuteCodeBlockActions<TLanguageKindEnum>(
             CancellationToken cancellationToken)
             where TLanguageKindEnum : struct
         {
-            ExecuteBlockActionsCore<CodeBlockStartAnalyzerAction<TLanguageKindEnum>, CodeBlockAnalyzerAction, SyntaxNodeAnalyzerAction<TLanguageKindEnum>, SyntaxNode, TLanguageKindEnum>(
-                codeBlockStartActions, codeBlockActions, codeBlockEndActions, analyzer,
-                declaredNode, declaredSymbol, executableCodeBlocks, (codeBlocks) => codeBlocks.SelectMany(
-                    cb =>
+            Debug.Assert(!executableCodeBlocks.IsEmpty);
+
+            // The actions we discover in 'addActions' and then execute in 'executeActions'.
+            var ephemeralActions = ArrayBuilder<SyntaxNodeAnalyzerAction<TLanguageKindEnum>>.GetInstance();
+            ExecuteBlockActionsCore(
+                startActions,
+                actions,
+                endActions,
+                analyzer,
+                declaredNode,
+                declaredSymbol,
+                operationBlocks: [],
+                semanticModel,
+                filterSpan,
+                isGeneratedCode,
+                addActions: static (startAction, endActions, args, cancellationToken) =>
+                {
+                    var (@this, startActions, analyzer, declaredNode, declaredSymbol, executableCodeBlocks, semanticModel, getKind, filterSpan, isGeneratedCode, ephemeralActions) = args;
+
+                    var scope = new HostCodeBlockStartAnalysisScope<TLanguageKindEnum>(startAction.Analyzer);
+                    var startContext = new AnalyzerCodeBlockStartAnalysisContext<TLanguageKindEnum>(
+                        scope, declaredNode, declaredSymbol, semanticModel, @this.AnalyzerOptions, filterSpan, isGeneratedCode, cancellationToken);
+
+                    // Catch Exception from the start action.
+                    @this.ExecuteAndCatchIfThrows(
+                        startAction.Analyzer,
+                        static args => args.startAction.Action(args.startContext),
+                        argument: (startAction, startContext),
+                        new AnalysisContextInfo(@this.Compilation, declaredSymbol, declaredNode),
+                        cancellationToken);
+
+                    endActions.AddAll(scope.CodeBlockEndActions);
+                    ephemeralActions.AddRange(scope.SyntaxNodeActions);
```

Comment:
note: i changed the logic here from teh original code this moved from.  IN the original code, these two lines were isnide the lambda passed to ExecuteAndCatchIfThrows.  This means we had to pass in SO MUCH MORE state into that lambda, and hten unpack it.  But we don't need that at all.  All we are protecting is the call out to user-code through `args.StartAction.Action(...)`.  Once that has returned, we just have raw data on the scope object we can process here outside of hte lambda.  this greatly simpified things.